examples/modernbert/modernbert.zig

```diff
@@ -0,0 +1,234 @@
+///
+/// The GeGLU activation function is a combination of the Gated Linear Unit (GLU) and the Gaussian Error Linear Unit (GeLU).
+///
+/// see: https://paperswithcode.com/method/geglu
+pub const ModernBertMLP = struct {
+    Wi: zml.nn.Linear,
+    Wo: zml.nn.Linear,
+
+    pub fn forward(self: ModernBertMLP, hidden_states: Tensor) Tensor {
+        // Perform Wi
+        const wi_output: Tensor = zml.call(self.Wi, .forward, .{hidden_states});
+
+        // Split into input and gate tensors along the last dimension
+        // input, gate = self.Wi(hidden_states).chunk(2, dim=-1)
+        const input = wi_output.slice1d(-1, .{ .end = intermediate_size });
+        const gate = wi_output.slice1d(-1, .{ .start = intermediate_size });
+
+        // Apply activation function to input and multiply by gate :
+        // self.Wo(self.drop(self.act(input) * gate))
+        const activated_input = input.gelu().mul(gate);
+
+        // Perform Wo
+        return zml.call(self.Wo, .forward, .{activated_input});
+    }
+};
+
+/// Performs multi-headed self attention on a batch of unpadded sequences.
+///
+/// TODO: If Flash Attention 2 is installed, this module uses Flash Attention to improve throughput.
+/// If Flash Attention 2 is not installed, the implementation will use SDPA,
+pub const ModernBertAttention = struct {
+    Wqkv: zml.nn.Linear,
+    Wo: zml.nn.Linear,
+    is_global_attention: bool = false,
+
+    /// sdpa_attention_forward
+    pub fn forward(
+        self: ModernBertAttention,
+        hidden_states: Tensor,
+        attention_mask: Tensor,
+        sliding_window_mask: Tensor,
+        position_ids: Tensor,
+    ) Tensor {
+        const batch_size = hidden_states.shape().dim(0);
+        const seq_length = hidden_states.shape().dim(1);
+        const num_heads = 12; // config.json: num_attention_heads
+        const head_dim = 64; // config.json hidden_size / num_attention_heads = 768 / 12
+
+        if (self.is_global_attention) {
+            log.info("Global attention", .{});
+        } else {
+            log.info("Local attention", .{});
+        }
+
+        // Project to query, key, value - { batch_size, seq_len, 3 * num_heads * head_dim }
+        var qkv: Tensor = zml.call(self.Wqkv, .forward, .{hidden_states}); // Wqkv.out.0
+
+        // Reshape to { batch_size, seq_len, 3, num_heads, head_dim }
+        qkv = qkv.reshape(.{ batch_size, seq_length, 3, num_heads, head_dim });
+        // Transpose for attention computation  { batch_size, seq_len, 3, num_heads, head_dim } -> { batch_size, num_heads, 3, seq_length, head_dim }
+        qkv = qkv.transpose(.{ 0, 3, 2, 1, 4 });
+
+        // Split into query, key, value tensors - each { batch_size, num_heads, seq_length, head_dim }
+        var q = qkv.slice1d(2, .{ .start = 0, .end = 1 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
+        var k = qkv.slice1d(2, .{ .start = 1, .end = 2 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
```

Comment:
i'd recommend to tag earlier in your function so you don't need the transpose(.{ 0, 3, 2, 1, 4 })